### ISSUE TITLE
Introduce FiringMode.Serial

### ISSUE THREAD

**Issue by gao-artur** [dotnet-state-machine/stateless#527]
Currently, stateless supports two `FiringMode`'s: `Immediate` and `Queued`. Both of them have their use cases, but unfortunately, neither of them matches my needs.
`Immediate` will change the state even if another state change is in progress by another thread hence can be used only in a single-threaded environment.
`Queued` will enqueue the state change and return to the caller immediately, while the enqueued job will be executed by the thread which executes the current state change. Both immediate return and forwarding the work to another thread are unacceptable in my scenario. In addition, it has its own multi-threaded usage problems.
I suggest introducing a new `FiringMode.Serial`. In this mode, the call will block if there is an active state change and will fire only after it is finished. I'm just starting to learn the source code of the library, but it looks like this can be achieved very easily by introducing a `SemaphoreSlim` based lock in the `InternalFire(Async)` methods.
```c#
private readonly SemaphoreSlim _firingLock = new SemaphoreSlim(initialCount: 1, maxCount: 1);

void InternalFire(TTrigger trigger, params object[] args)
{
    switch (_firingMode)
    {
        ...
        case FiringMode.Serial:
            _firingLock.Wait();
            try
            {
                InternalFireOne(trigger, args);
            }
            finally
            {
                _firingLock.Release();
            }
            break;
        default:
           ...
    }
}

async Task InternalFireAsync(TTrigger trigger, params object[] args)
{
    switch (_firingMode)
    {
        ...
        case FiringMode.Serial:
            await _firingLock.WaitAsync();
            try
            {
                await InternalFireOneAsync(trigger, args);
            }
            finally
            {
                _firingLock.Release();
            }
            break;
        default:
            ...
    }
}
```

Also, since the lock can be held for a long time, it's good to introduce `Fire(Async)` overloads that accept `CancellationToken`.

**Comment by leeoades** [comment]
I didn't think stateless supported being used in a multithreaded way. 

Having multiple threads attempting state changes seems extremely problematic. One thread could be checking a guard whilst another thread changes the state immediately. The first thread then approves the state change and moves to the target state which may not be a permittable transition.

I wouldn't be encouraging users to interact with a state machine in a multithreaded fashion. If I were you, I would be looking to queue the transitions and have them processed sequentially.

**Comment by gao-artur** [comment]
You are right, stateless is not thread-safe, and it's documented as such. But the fact the `FiringMode.Queued` exists suggests it can be used in a multithreaded environment (even though there is a race condition on the [_firing](https://github.com/dotnet-state-machine/stateless/blob/dev/src/Stateless/StateMachine.cs#L361-L387) field).
The new mode I suggest will execute all `Fire` methods serially, which means concurrent callers will be blocked until the active transition completes. This is similar to enqueueing transitions with the important difference: the caller won't be released before the transition it initialized is finished (and this is a must requirement in my use case). Hence this scenario is not possible:

> One thread could be checking a guard whilst another thread changes the state immediately. The first thread then approves the state change and moves to the target state which may not be a permittable transition.

As a workaround, I'm currently using the `SemaphoreSlim` lock on the public methods that `Fire`. But built-in support sounds to me legit and less fragile.

**Comment by gao-artur** [comment]
There is a caveat with my proposal: if the entry action of one state `Fire`s internally it will cause a deadlock because `SemaphoreSlim` doesn't support re-entrance...

**Comment by leeoades** [comment]
I would always wrap anything in a multithreaded environment to de-parallelise as soon as possible (wherever that makes sense) at a boundary so that everything within can be safely single threaded.
I would do the same here. If you must have your threads block and fire triggers sequentially then have them controlled by a semaphore slim in a layer above stateless rather than trying to bring that into the statemachine.

**Comment by gao-artur** [comment]
Right, the workaround exists. The problem with any workaround is that it must be remembered every time a change is introduced in the code. A new developer who will maintain the code after me, or even me, must remember to wrap the call with a semaphore. The question is if the built-in support is legitimate. I think it is, but with a different implementation to avoid deadlocks.

**Comment by crozone** [comment]
I think there is some misconception about the way in which the firing modes are designed to work, as well as their intended use-cases. Perhaps this needs to be cleaned up in the documentation.

> But the fact the FiringMode.Queued exists suggests it can be used in a multithreaded environment (even though there is a race condition on the [_firing](https://github.com/dotnet-state-machine/stateless/blob/dev/src/Stateless/StateMachine.cs#L361-L387) field).

Stateless is not designed to handle any concurrency. I'd like to avoid using the term "mulithreaded", because it is confusing when discussing Stateless' async methods. A state machine object is fine to be called from many different threads, so long as the methods are not called *concurrently*. For example the execution flow might bounce between many different threadpool threads when calling Stateless' async methods, but they're still only being run in a serialized fashion, one after the other.

Back to `FiringMode.Queued`, it is not designed for concurrency either. Rather, it is specifically intended to support the use-case where you call a `Fire()` method from within a transition function itself. This enqueues the firing of the trigger until all of the currently executing transition functions have completed. This is often the expected behaviour, which is why `FiringMode.Queued` is the default.

Consider the following triggers:

```
stateMachine.Configure(State.StateA)
.Permit(Trigger.StateBTransitionTrigger, State.StateB)
.OnEntryAsync(async () =>
{
    Console.WriteLine("A");
    await stateMachine.FireAsync(Triggers.StateBTransitionTrigger);
    Console.WriteLine("B");
});

stateMachine.Configure(State.StateB)
.OnEntryAsync(async () =>
{
    Console.WriteLine("C");
});
```

With `FiringMode.Immediate`, the call to `stateMachine.FireAsync(Triggers.StateBTransitionTrigger)` will immediately start executing all of the transition functions for `State.StateB` and then transition the state into StateB *during the fire call*. Then the fire call will return, and the rest of the lambda will execute. The console would print A, C, B.

With `FiringMode.Queued`, the call to `stateMachine.FireAsync(Triggers.StateBTransitionTrigger)` will be delayed until after all of the current state transition functions have completed. The console would print A, B, C.

It sounds like what you want with `FiringMode.Serial` is actually unrelated to these firing modes - instead, you need some sort of work queue which will serialize the firing of triggers coming in from several concurrent sources. This work queue can then dequeue the trigger events and call the state machine in a serialized fashion.

This is similar to the "do activities" issue here: https://github.com/dotnet-state-machine/stateless/issues/77

See my other comment here for an example implementation of a work queue which can serialize the state machine fire events: https://github.com/dotnet-state-machine/stateless/issues/77#issuecomment-1475641964

**Comment by gao-artur** [comment]
@crozone, thanks for the explanation of the `FiringMode.Queued`. I didn't even think in this direction. The documentation around this mode definitely should be improved. But I have to admit I don't have experience with FSM's. Maybe it's something trivial in this world.

Currently, `SemaphoreSlim` works for me because it preserves the order of waiting tasks, so it's similar to using a firing queue.
I thought about queues but for a different reason. In addition to what I described above, I need to give a higher priority for some transitions (for example, `Abort` transition should be executed immediately after the active transition, no matter how many are already waiting in the queue). 

Continuing the idea of the worker task and priorities when using the new mode (you can call it `Serial` or whatever sounds more correct), the worker task can be created and start polling tasks from the queue. Any usual transitions should have "Normal" priority, a transition triggered from another transition should have "Highest", and the user should be able to define "High" priority transitions to execute before "Normal" (names are just for example, numbers can be used for more flexibility).

I think this way, thread safety can be built-in into the library while providing additional functionality.

Note that `Channel` doesn't support priorities, and the new [PriorityQueue](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.priorityqueue-2?view=net-7.0) is not stable (doesn't preserve the order within priority).

**Comment by crozone** [comment]
@mclift Do you think we should start a proper design discussion for a `FiringMode.Serial`, given that we keep getting issues relating to concurrent queuing of triggers? It's definitely a super common usecase.

I think the work items would be:

1. More clearly document the differences between `FiringMode.Immediate` and `FiringMode.Queued`, with an explicit explanation that neither are threadsafe.
2. Design a `FiringMode.Serial / Serialized` mode which internally queues triggers in a threadsafe manner.
3. Put together a draft PR. I now agree with @gao-artur that `SemaphoreSlim` is likely ideal for this since it handles synchronous and asynchronous synchronization simultaneously, with the difficult code handled by the framework.

**Comment by crozone** [comment]
I've been looking into this and considering the use of `SemaphoreSlim`. While researching the implementation of `AsyncLock`. I've come to the conclusion that we actually don't even necessarily need a `SemaphoreSlim`.

Since we're simply serializing the firing of triggers, we can get away with as something as simple as a `ContinueWith()` chain:

```csharp
private object currentTaskLock = new object();
private Task currentTask = Task.CompletedTask;
private Task EnqueueTask(Func<Task> action)
{
    lock (currentTaskLock)
    {
        currentTask = currentTask.IsCompleted
            ? action()
            : currentTask.ContinueWith((t) => action(),
                CancellationToken.None,
                TaskContinuationOptions.ExecuteSynchronously,
                TaskScheduler.Default);

        return currentTask;
    }
}
```

The basic idea is that any currently executing Fire() task is stored as the `currentTask`. If a new task is queued while the current task is still in progress, it is simply queued as a continuation of the current task, and the current task is replaced with the continuation task. If the current task has finished, it is instead replaced with the new task to avoid storing a huge chain of completed tasks.

**Comment by gao-artur** [comment]
I'm not an expert on `ContinueWith` methods, so that I may be wrong. But according to the docs, the `TaskContinuationOptions.ExecuteSynchronously` will run the continuation state change on the same thread as the previous state change. It means the previous state change caller will be unnecessarily blocked for a prolonged time. I think you should use `TaskContinuationOptions.RunContinuationsAsynchronously` here.

**Comment by leeoades** [comment]
Hey Guys,

Making anything thread-safe is no trivial task. Plus it has performance
implications for all single-threaded usages plus increased maintenance
complexity going forward - as developers need to ensure they do not
introduce a new possible vector for corruption in a
multi-threaded environment.

I've recently spent some effort to ensure the synchronization context is
not lost so that the state machine can be used within Microsoft Orleans.
Any introduction of continuations etc needs to ensure they do not undo this.

Please may I ask what the use cases are that we are trying to solve? In my
experience, synchronising threads should be done well before it would reach
a component such as our state machine.

Lee.

**Comment by crozone** [comment]
@leeoades I appreciate your concern, stateless has very large consumers that we don't want to subtly break with a change like this.

The use-case for `FiringMode.Serial` is where a state machine is driven asynchronously by several other tasks which each generate events (eg tasks reading from serial ports, receiving messages on sockets, etc). All of these events are fed into the state machine as triggers which can change application state. This seems to be a common scenario for many smaller embedded style applications where the state machine forms the core of the application logic.

Since the state machine is not threadsafe, the trigger needs to be first put into some sort of work queue (like a `Channel<T>`) and then processed on a dedicated worker thread. This works nicely, but is additional boilerplate that needs to be re-implemented for a seemingly common use-case.

The state machine already contains an internal queuing mechanism for the case when triggers are fired from within a state transition, which is enabled when the firing mode is set to `FiringMode.Queued` (the default). If this internal queuing mechanism could be made threadsafe, it can effectively be used as the work queue. The semantics of how this works needs to be carefully considered to avoid deadlocks (eg when firing triggers from within a transition), but it should theoretically work quite well.

This would be implemented as an additional firing mode, `FiringModel.Serial`, with its own dedicated codepath. `FiringMode.Immediate` and `FiringMode.Queued` already have their own codepaths, so the code is already structured this way.

The default firing mode, `FiringMode.Queued`, would remain as it is (not threadsafe) and should/must not suffer a performance regression as a result of this feature. It should ideally remain almost entirely untouched. We definitely need to unsure the synchronization context is maintained (I believe we have a unit test for this?).

In any case we are definitely not going to rush into implementing this without plenty of prototyping. It may turn out to not be a good idea.

Ryan.

**Comment by leeoades** [comment]
Thanks for taking the time to respond, @crozone 

If this were an event-driven scenario - where the events come in, adjust the state and cause change events - then I think queuing the state changes would work. This is essentially a read model.

Funnily enough, I built an RX State Machine a looong time ago. In this case, RX itself was responsible for synchronising the calls, and there was no ability to read the state machine. The events coming in only led to events going out. 

Any interrogation of the current state by the consumer is unreliable and could lead to unintended behaviour since of course the current state could be behind with queued triggers yet to be processed. The consumer might raise a trigger thinking they were affecting the state in one way but instead they are queuing the trigger and they act against a potentially different state once the trigger is processed. 

If the entire interaction with the state machine was atomic - it could interrogate the current state when it was that consumer's turn and triggers fired would occur immediately. Queuing these interactions at this higher level is safer and more intuitive. Synchronise all traffic through not only the state machine but through the wider context.

Why not fork the repo and build your multi-threaded version? I'll be happy to throw some multithreaded unit tests at it.

**Comment by jmo-gelsight** [comment]
Even after reading this discussion I am still unclear on one point.  

If I use the `StateMachine` from multiple threads (with `FiringMode.Queued`) but I protect every single call to the `Fire()` function with a simple lock, eg.

    lock(_myLock)
        Machine.Fire(Trigger.DoSomething);
        
  Is *that* thread safe?

**Comment by tedchirvasiu** [comment]
Until this gets settled one way or the other, for people looking for a solution without having to create a fork or creating an entire decorator, I'm thinking about an extension method like:

```csharp
using System.Runtime.CompilerServices;

namespace Stateless.Threadsafe.Extensions {
    public static class StatelessExtensions {

        class SyncContext {
            public Task CurrentTask = Task.CompletedTask;
        }

        static ConditionalWeakTable<object, SyncContext> SyncTable = new ConditionalWeakTable<object, SyncContext>();

        public static Task FireThreadsafeAsync<TState, TTrigger>(
            this StateMachine<TState, TTrigger> stateMachine,
            TTrigger trigger,
            params object[] args
        ) {

            var context = SyncTable.GetOrCreateValue(stateMachine);

            lock (context) {
                context.CurrentTask = context.CurrentTask.IsCompleted ?
                    stateMachine.FireAsync(trigger, args) :
                    context.CurrentTask.ContinueWith(
                        (t) => stateMachine.FireAsync(trigger, args),
                        CancellationToken.None,
                        TaskContinuationOptions.ExecuteSynchronously,
                        TaskScheduler.Default
                    )
                    .Unwrap();

                return context.CurrentTask;
            }
        }
    }
}
```

This is inspired by @crozone's proposal above with ContinueWith, but it uses a [ConditionalWeakTable](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2) to externally hold the currently executing task for each StateMachine.
This however means that you must be careful and use `FireThreadsafeAsync` everywhere instead of the default `FireAsync` (otherwise you might still get race conditions).

If you need the other overloads for FireAsync, you can easily add them just like they are done in the source code: https://github.com/dotnet-state-machine/stateless/blob/523efd6225cf60b0eb85f0862cdd006a47962da8/src/Stateless/StateMachine.Async.cs#L101-L106

Warning, I have not yet thoroughly tested this extension method.

Edit: 14.12.2025 - Added `TaskContinuationOptions.ExecuteSynchronously` just like in @crozone's example, without it calling `FireThreadsafeAsync` from within an OnEntryAsync will cause a deadlock.

**Comment by leeoades** [comment]
I'll be honest - I'm struggling to understand why you would want to allow more than one thread to be changing the state machine concurrently? I fear I'm missing the point.
I would expect all triggers to be serialised - with an entire transition to have completed, including all resulting action executions - before the next trigger was allowed entry.
If a resulting execution should not be awaited, then the call to it could signal its start and return immediately.
Perhaps someone could kindly enlighten me? Many thanks,

**Comment by tedchirvasiu** [comment]
@leeoades I think @crozone described it pretty well in his comment above:

> The use-case for FiringMode.Serial is where a state machine is driven asynchronously by several other tasks which each generate events (eg tasks reading from serial ports, receiving messages on sockets, etc). All of these events are fed into the state machine as triggers which can change application state. This seems to be a common scenario for many smaller embedded style applications where the state machine forms the core of the application logic.
> 
> Since the state machine is not threadsafe, the trigger needs to be first put into some sort of work queue (like a Channel<T>) and then processed on a dedicated worker thread. This works nicely, but is additional boilerplate that needs to be re-implemented for a seemingly common use-case.

Most of us wouldn't want necessarily to mutate the State Machine itself from multiple threads, just be able to queue events onto it in a thread safe manner. The events should still be processed sequentially.

I think most of us who need this behavior would not expect `FiringMode.Immediate` to work in such scenarios, but would expect something like `FiringMode.Queued` to be safe. I guess hence @gao-artur's proposal for `FiringMode.Serial` (possibly in order not to break existing code).

Ensuring yourself all calls to the State Machine are serialized kinda sucks because:
1. If you do it naively with a semaphore, you will deadlock if you call the trigger from OnEntry/OnEntryAsync etc.
2. If you do it with a `Channel` or a `ConcurrentQueue` or anything like that, you're pretty much replicating/duplicating Stateless' already existing queueing mechanism and as @crozone said "it is additional boilerplate that needs to be re-implemented for a seemingly common use-case"
3. If you wanna ensure nobody can call the unsafe Fire method from the vanilla state machine, you have to make a decorator / facade which would force you to redeclare all the functions exposed by the State Machine. Fire/FireAsync are not virtual, so inheritance does not help since you would have to either create a FireThreadSafeAsync variant or an overload for the original Fire/FireAsync and developers could still call the unsafe version.

**Comment by leeoades** [comment]
@tedchirvasiu Oh, I think I see. So, the serialising of the triggers in whatever way would occur - it's just that we could do this internally. I'm not sure I would want to support this because it feels like we'd be encouraging it! 

So, my thoughts are:
- if we used a channel / queue then the triggering thread would return immediately and we would need to make it clear that it has only been queued.
  - and as you stated, to block the triggering thread is asking for deadlocks.
- the queue of triggers is not persisted in any form so could be lost if the process shut down.
- the trigger of the state machine is unlikely to be the only operation in which case other code along side it would need to be threadsafe - and I always discourage this sort of code.

**Comment by tedchirvasiu** [comment]
@leeoades Yeah, it would mostly be a thing of convenience.

> if we used a channel / queue then the triggering thread would return immediately and we would need to make it clear that it has only been queued.

The thing is, this is pretty much how `FiringMode.Queued` seems to work already:

https://github.com/dotnet-state-machine/stateless/blob/523efd6225cf60b0eb85f0862cdd006a47962da8/src/Stateless/StateMachine.Async.cs#L177-L201

If it's already running, it adds the trigger to the queue and returns. So the only thing that seems unsafe is this block:
```csharp
            if (_firing)
            {
                _eventQueue.Enqueue(new QueuedTrigger { Trigger = trigger, Args = args });
                return;
            }
```

paired with

```csharp
var queuedEvent = _eventQueue.Dequeue();
```

Because flipping the `_firing` boolean and `_eventQueue.Enqueue()` and `_eventQueue.Dequeue()` are not thread safe. And it feels like making this chunk thread-safe could be pretty straight forward. I'm thinking an elaborate way would be to use [Interlocked.CompareExchange](https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked?view=net-10.0) to flip the boolean (which would have to be an int in this case) and a [ConcurrentQueue](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentqueue-1?view=net-10.0), however a simple lock might be more than enough and possibly even more efficient.

```csharp
lock(this) {
    if (_firing) {
        _eventQueue.Enqueue(new QueuedTrigger { Trigger = trigger, Args = args });
         return;
    }

     _firing = true;
}
```

So just wrap the enqueueing in a lock and put the `_firing = true;` inside the lock instead of the try-finally block. This would ensure only one thread is allowed to process the triggers at a time.

Then a lock would be needed for Dequeue-ing too:

```csharp
while (true) {

   QueuedTrigger queuedEvent 

   lock(this) {
        if (_eventQueue.Count == 0)
            break;
         
        queuedEvent = _eventQueue.Dequeue();           
   }
   
   await InternalFireOneAsync(queuedEvent.Trigger, queuedEvent.Args).ConfigureAwait(RetainSynchronizationContext);
}
```

And basically the same thing on the Fire (sync) method too.
Of course this might impact existing consumers relying on nanosecond performance.

**Comment by leeoades** [comment]
Another question is, should a transition that occurs as a direct result of another transition take precedence over an already queued trigger? For example, if Trigger Goto A and Trigger Goto B are fired in quick succession, Goto A is being processed and Goto B is queued. If in the processing of Goto A, there is a Trigger Goto Z, then should that occur before attempting to go to B? I, personally, think it should - otherwise all sorts of integrity seems to be lost. In which case, there would need to be an inner queue or priority queue perhaps.

**Comment by tedchirvasiu** [comment]
@leeoades Hmm, this is a good question. In the current `FiringMode.Queued` it does not, it is simply added to the end of the queue (assuming `Goto B` got fired before `Goto Z`). So I think the result would be `A, B, Z` (since Z would return early as shown in the block above).

However, in that case since there is no parallel code, it would definitely be more obvious and predictable that you called `Goto B` sometime before `Goto Z` (since it would have to be somewhere in the OnEntry/OnEntryAsync as well)

In my implementation of `FiringMode.Serial` ( #639 ) if you are currently processing trigger `Goto A` and while it is being executed, another parallel thread enqueues `Goto B`, then if in the processing of `Goto A` there is a trigger `Goto Z`, at the moment of firing `Goto Z` you might not know or expect `Goto B` to have sneaked in, but it definitely can (since I don't lock the entire state machine down while it is processing).

I am not sure how one would implement a queue of that sort because you would need to know whether a trigger got Fired from within the state machine or from the outside (and everyone calls the same Fire method).

The other option is to lock the entire state machine down while it is firing, however that will basically certainly deadlock if you call Trigger from OnEntry/OnEntryAsync. We would need a very special re-entrant lock which would permit re-entry if Fire got called from within the state machine (which is the same conundrum as at the paragraph above). Something with `AsyncLocal` might work, but it could also be very hacky or bug-prone.

My personal take is that if your state machine has to endure multi-threading, you have to prepare it to accept invalid / stale triggers at any-time. So you have to assume your SM might get bombarded with events which at the time of processing might be stale (because at the time they got fired the SM was in a different older state). And this includes events coming from within the SM.

So basically ignore unexpected Triggers or handle invalid trigger for state X exceptions.
